### DISCUSSION TITLE
node-fetch is not working with App dir

### DISCUSSION THREAD

**Discussion by SergeySypalo** [vercel/next.js#53200] [Help] [answered]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
I want to fetch BTC price from Binance using CCXT library. The code works when I use pages dir, but doesn't when I use /app dir
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router

### Link to the code that reproduces this issue or a replay of the bug

n/a

### To Reproduce

To reproduce the bug:
1) create a new next app
2) run "npm install ccxt"
3) replace te code in main page to:

import ccxt from 'ccxt'

async function GetData() {	
	const exchange = new ccxt.binance()
	const price = await exchange.fetchTotalBalance()
	return price
}

export default async function Page() {
	const price = await GetData()
	return price.last
}

### Describe the Bug

Unhandled Runtime Error
Error: Cannot find module 'D:\repos\portfolio-manager\.next\server\static_dependencies\node-fetch\index.js' imported from D:\repos\portfolio-manager\.next\server\app\page.js

### Expected Behavior

Last Bitcoin price should be displayed

**Comment by miketeix** [comment]
I have the same issue for node-fetch w/ CCXT. Also it's not finding a couple of modules in the 'ws' library. 

This goes away when I use node externals with my webpack config (inside of next.config):
```
/** @type {import('next').NextConfig} */
const nodeExternals = require("webpack-node-externals")
const nextConfig = {
    webpack: (config, { isServer }) => {
        // eslint-disable-next-line space-before-blocks
        if (isServer) {
            config.externals = nodeExternals()
        }
        return config
    }
}
module.exports = nextConfig
```

but then webpack doesn't seem to package correctly for the client side despite using the `isServer` flag. E.g. `Error: Cannot find module 'next/font/google/target.css?...`

Any clues here?

**Comment by herikaniugu** [comment]
anyone who know how to fix it?

**Reply by SergeySypalo** [reply]
I switched back to /pages for now
As no real fix has been released(

**Comment by SergeySypalo** [comment]
Just tested with newer versions, but error is still present:
node: 20.5.1
npm: 9.8.1
nextjs: 13.4.19
react: 18.2
ccxt: 4.0.77

**Comment by bludnic** [comment, marked as answer]
Solved by replacing fetch implementation of CCXT with the NextJS `fetch`.

```tsx
import ccxt from 'ccxt'

const fetcher = (...args: Parameters<typeof fetch>) => fetch(...args);

export default async function Page() {
  const exchange = new ccxt.binance()
  exchange.fetchImplementation = fetcher // <-

  const price = await exchange.fetchTicker('BTC/USDT')

  return <div>{JSON.stringify(price)}</div>
}
```

**Reply by SergeySypalo** [reply]
Thanks, that worked!
